File: algorithms/2_medium/D/00740_delete_and_earn/dp_for_any_vals.cpp
class Solution {
public:
    int deleteAndEarn(vector<int>& nums) {
        unordered_map<int, int> freqs{};

        for (const auto& num : nums) {
            freqs[num] += 1;
        }

        vector<pair<int, int>> dp;
        const size_t dp_len{ freqs.size() };
        dp.reserve(dp_len);

        for (const auto& [val, freq] : freqs) {
            dp.push_back({ val, val * freq });
        }

        ranges::sort(dp, [](const auto& a, const auto& b) {
            return a.first < b.first;
        });

        if (dp_len == 1) {
            return dp[0].second;
        }

        if (dp[0].first + 1 < dp[1].first) {
            dp[1].second += dp[0].second;
        }

        if (dp_len == 2) {
            return std::max(dp[0].second, dp[1].second);
        }

        if (dp[1].first + 1 < dp[2].first) {
            dp[2].second += std::max(dp[0].second, dp[1].second);
        } else {
            dp[2].second += dp[0].second;
        }

        for (size_t i{ 3 }; i < dp_len; ++i) {
            if (dp[i - 1].first + 1 < dp[i].first) {
                dp[i].second += std::max(dp[i - 1].second, dp[i - 2].second);
            } else {
                dp[i].second += std::max(dp[i - 2].second, dp[i - 3].second);
            }
        }

        return std::max(dp[dp_len - 1].second, dp[dp_len - 2].second);
    }
};

File: algorithms/2_medium/D/00740_delete_and_earn/dp_for_any_vals.rs
use std::collections::HashMap;

impl Solution {
    pub fn delete_and_earn(nums: Vec<i32>) -> i32 {
        let mut freqs: HashMap<i32, i32> = HashMap::new();

        for num in nums.into_iter() {
            *freqs.entry(num).or_insert(0) += 1;
        }

        let mut dp = freqs.into_iter()
                          .map(|(val, freq)| (val, val * freq))
                          .collect::<Vec<(i32, i32)>>();

        dp.sort_unstable();
        let dp_len = dp.len();
        
        if dp_len == 1 {
            return dp[0].1;
        }

        dp[1].1 += if dp[0].0 + 1 < dp[1].0 { dp[0].1 } else { 0 };

        if dp_len == 2 {
            return dp[0].1.max(dp[1].1);
        }

        dp[2].1 += if dp[1].0 + 1 < dp[2].0 { dp[0].1.max(dp[1].1) } else { dp[0].1 };

        for i in 3..dp_len {
            dp[i].1 += {
                if dp[i - 1].0 + 1 < dp[i].0 { 
                    dp[i - 1].1.max(dp[i - 2].1)
                } else {
                    dp[i - 2].1.max(dp[i - 3].1)
                }
            };
        }

        dp[dp_len - 1].1.max(dp[dp_len - 2].1)
    }
}
